MapleServer2/PacketHandlers/Game/ItemExchangeHandler.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Maple2Storage.Types.Metadata;
+using MaplePacketLib2.Tools;
+using MapleServer2.Constants;
+using MapleServer2.Data.Static;
+using MapleServer2.Packets;
+using MapleServer2.Servers.Game;
+using MapleServer2.Tools;
+using MapleServer2.Types;
+using Microsoft.Extensions.Logging;
+
+namespace MapleServer2.PacketHandlers.Game
+{
+    public class ItemExchangeHandler : GamePacketHandler
+    {
+        public override RecvOp OpCode => RecvOp.ITEM_EXCHANGE;
+
+        public ItemExchangeHandler(ILogger<ItemExchangeHandler> logger) : base(logger) { }
+
+        private enum ItemExchangeMode : byte
+        {
+            Use = 0x1,
+        }
+
+        public override void Handle(GameSession session, PacketReader packet)
+        {
+            ItemExchangeMode mode = (ItemExchangeMode) packet.ReadByte();
+
+            switch (mode)
+            {
+                case ItemExchangeMode.Use:
+                    HandleUse(session, packet);
+                    break;
+                default:
+                    IPacketHandler<GameSession>.LogUnknownMode(mode);
+                    break;
+            }
+        }
+
+        public enum ExchangeNotice : short
+        {
+            Sucess = 0x0,
+            Invalid = 0x1,
+            CannotFuse = 0x2,
+            InsufficientMeso = 0x3,
+            InsufficientItems = 0x4,
+            EnchantLevelTooHigh = 0x5,
+            ItemIsLocked = 0x6,
+            CheckFusionAmount = 0x7,
+        }
+
+        private static void HandleUse(GameSession session, PacketReader packet)
+        {
+            long itemUid = packet.ReadLong();
+            long unk = packet.ReadLong();
+            int quantity = packet.ReadInt();
+
+            if (!session.Player.Inventory.Items.ContainsKey(itemUid))
+            {
+                return;
+            }
+
+            Item item = session.Player.Inventory.Items[itemUid];
+
+            ItemExchangeScrollMetadata exchange = ItemExchangeScrollMetadataStorage.GetMetadata(item.FunctionId);
+
+            if (!session.Player.Wallet.Meso.Modify(-exchange.MesoCost * quantity))
+            {
+                session.Send(ItemExchangePacket.Notice((short) ExchangeNotice.InsufficientMeso));
+                return;
+            }
+
+            if (exchange.ItemCost.Count != 0)
+            {
+                if (!PlayerHasAllIngredients(session, exchange, quantity))
+                {
+                    session.Send(ItemExchangePacket.Notice((short) ExchangeNotice.InsufficientItems));
+                    return;
+                }
+
+                if (RemoveRequiredItemsFromInventory(session, exchange, item, quantity))
+                {
+                    Item exchangeRewardItem = new(exchange.RewardId)
```

Comment:
84-91 matches 96-104 almost exactly, maybe do
```InventoryController.Consume(session, item.Uid, exchange.RecipeAmount * quantity);``` in that else, then the rest can be the code where they are identical